File: backend/src/main/java/org/sport/backend/dto/base/PageResponse.java
package org.sport.backend.dto.base;

import lombok.*;
import org.springframework.data.domain.Page;

import java.util.Collections;
import java.util.List;

@Builder
@NoArgsConstructor
@AllArgsConstructor
@Getter
@Setter
public class PageResponse<T> {
    private int currentPage;
    private int totalPages;
    private int pageSize;
    private long totalElements;
    @Builder.Default
    private List<T> data= Collections.emptyList();

    public static <T> PageResponse<T> of(Page<?> page, List<T> data) {
        return PageResponse.<T>builder()
                .currentPage(page.getNumber() + 1)
                .pageSize(page.getSize())
                .totalPages(page.getTotalPages())
                .totalElements(page.getTotalElements())
                .data(data)
                .build();
    }
}

File: mobile/lib/models/page_response.dart
class PageResponse<T> {
  final int currentPage;
  final int totalPages;
  final int pageSize;
  final int totalElements;
  final List<T> data;

  PageResponse({
    required this.currentPage,
    required this.totalPages,
    required this.pageSize,
    required this.totalElements,
    required this.data,
  });

  factory PageResponse.fromJson(Map<String, dynamic> json, T Function(dynamic) fromJsonT) {
    return PageResponse(
      currentPage: json['currentPage'] ?? 0,
      totalPages: json['totalPages'] ?? 0,
      pageSize: json['pageSize'] ?? 0,
      totalElements: json['totalElements'] ?? 0,
      data: (json['data'] as List).map((i) => fromJsonT(i)).toList(),
    );
  }
}